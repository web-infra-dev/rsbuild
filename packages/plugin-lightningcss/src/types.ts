import type lightningcss from 'lightningcss';
import type { CustomAtRules, TransformOptions } from 'lightningcss';

export type Lightningcss = typeof lightningcss;

export type LightningCSSTransformOptions = Omit<
  TransformOptions<CustomAtRules>,
  'filename' | 'code' | 'inputSourceMap'
>;

export type LightningCSSLoaderOptions = LightningCSSTransformOptions & {
  implementation?: Lightningcss;
};

export type PluginLightningcssOptions = {
  /**
   * @see https://github.com/parcel-bundler/lightningcss/blob/master/node/index.d.ts
   * @default
   * {
   *   targets: browserslistToTargets(browserslist)
   * }
   */
  transform?: boolean | LightningCSSTransformOptions;
  /**
   * lightningcss instance
   * @example
   * import { pluginLightningcss } from '@rsbuild/plugin-lightningcss';
   * import * as lightningcss from 'lightningcss';
   *
   * pluginLightningcss({
   *    implementation: lightningcss,
   * })
   */
  implementation?: unknown; // loose type of `typeof import('lightningcss')`
};
