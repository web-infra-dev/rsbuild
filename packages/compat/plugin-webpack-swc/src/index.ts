export { default as SwcLoader } from './loader';
export { pluginSwc } from './plugin';
export * from './binding';
export * from './types';
