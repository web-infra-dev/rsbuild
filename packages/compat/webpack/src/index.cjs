/**
 * CommonJS wrapper
 */
module.exports.webpackProvider = (...args) =>
  import('./index.js').then((i) => i.webpackProvider(...args));
