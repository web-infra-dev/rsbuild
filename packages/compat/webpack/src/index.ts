export { webpackProvider } from './provider';
export type { WebpackProvider } from './provider';
export { webpackBuild } from './core/build';
export { createDefaultConfig } from './config/defaults';
export type {
  LessLoaderOptions,
  SassLoaderOptions,
  PostCSSLoaderOptions,
  CSSLoaderOptions,
  StyleLoaderOptions,
} from '@rsbuild/shared';
export type {
  RsbuildPlugin,
  RsbuildPluginAPI,

  // Config Types
  DevConfig,
  HtmlConfig,
  ToolsConfig,
  OutputConfig,
  SourceConfig,
  RsbuildConfig,
  SecurityConfig,
  NormalizedConfig,
  PerformanceConfig,
  ExperimentsConfig,

  // Hook Callback Types
  ModifyWebpackChainFn,
  ModifyWebpackConfigFn,
  ModifyWebpackChainUtils,
  ModifyWebpackConfigUtils,

  // Third Party Types
  webpack,
  WebpackChain,
  WebpackConfig,
  CSSExtractOptions,
  HTMLPluginOptions,
} from './types';
