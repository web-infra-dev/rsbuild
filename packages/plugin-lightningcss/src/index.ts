export { pluginLightningcss } from './plugin';

import * as lightningcss from 'lightningcss';
export { lightningcss };
