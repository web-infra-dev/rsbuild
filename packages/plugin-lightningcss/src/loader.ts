/**
 * modified from https://github.com/fz6m/lightningcss-loader
 * MIT License https://github.com/fz6m/lightningcss-loader/blob/main/LICENSE
 * Author @fz6m
 */
import type { LoaderContext } from '@rspack/core';
import { transform as _transform } from 'lightningcss';
import { Buffer } from 'node:buffer';
import type { LightningCSSLoaderOptions } from './types';

const LOADER_NAME = 'lightningcss-loader';

async function LightningCSSLoader(
  this: LoaderContext<LightningCSSLoaderOptions>,
  source: string,
  prevMap?: Record<string, any>,
): Promise<void> {
  const done = this.async();
  const options = this.getOptions();
  const { implementation, targets, ...opts } = options;

  if (
    implementation &&
    typeof (implementation as any).transform !== 'function'
  ) {
    done(
      new TypeError(
        `[${LOADER_NAME}]: options.implementation.transform must be an 'lightningcss' transform function. Received ${typeof (implementation as any).transform}`,
      ),
    );
    return;
  }

  const transform: typeof _transform =
    (implementation as any)?.transform ?? _transform;

  try {
    const { code, map } = transform({
      filename: this.resourcePath,
      code: Buffer.from(source),
      sourceMap: this.sourceMap,
      targets,
      inputSourceMap:
        this.sourceMap && prevMap ? JSON.stringify(prevMap) : undefined,
      ...opts,
    });
    const codeAsString = code.toString();
    done(null, codeAsString, map && JSON.parse(map.toString()));
  } catch (error: unknown) {
    done(error as Error);
  }
}

export default LightningCSSLoader;
