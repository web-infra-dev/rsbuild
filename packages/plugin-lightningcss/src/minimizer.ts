import { Buffer } from 'node:buffer';
import { CSS_REGEX } from '@rsbuild/shared';
import type { Rspack } from '@rsbuild/shared';
/**
 * modified from https://github.com/fz6m/lightningcss-loader
 * MIT License https://github.com/fz6m/lightningcss-loader/blob/main/LICENSE
 * Author @fz6m
 */
import { transform as _transform } from 'lightningcss';
import type { LightningCSSMinifyPluginOptions } from './types';

const PLUGIN_NAME = 'lightningcss-minify-plugin';

export class LightningCSSMinifyPlugin {
  private readonly options: LightningCSSMinifyPluginOptions;

  private readonly transform: typeof _transform;

  constructor(opts: LightningCSSMinifyPluginOptions = {}) {
    const { implementation } = opts;

    if (
      implementation &&
      typeof (implementation as any).transform !== 'function'
    ) {
      throw new TypeError(
        `[${PLUGIN_NAME}]: implementation.transform must be an 'lightningcss' transform function. Received ${typeof (implementation as any).transform}`,
      );
    }

    this.transform = (implementation as any)?.transform ?? _transform;
    this.options = opts;
  }

  apply(compiler: Rspack.Compiler) {
    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
      compilation.hooks.processAssets.tapPromise(
        {
          name: PLUGIN_NAME,
          stage: (compilation as any)?.PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE,
        },
        async () => await this.transformAssets(compilation),
      );

      compilation.hooks.statsPrinter.tap(PLUGIN_NAME, (statsPrinter) => {
        statsPrinter.hooks.print
          .for('asset.info.minimized')
          .tap(PLUGIN_NAME, (minimized, { green, formatFlag }) =>
            minimized && green && formatFlag
              ? green(formatFlag('minimized'))
              : '',
          );
      });
    });
  }

  private async transformAssets(
    compilation: Rspack.Compilation,
  ): Promise<void> {
    const {
      options: { devtool },
      webpack: {
        sources: { SourceMapSource, RawSource },
      },
    } = compilation.compiler;

    const sourcemap =
      this.options.sourceMap === undefined
        ? ((devtool && (devtool as string).includes('source-map')) as boolean)
        : this.options.sourceMap;

    const { targets: userTargets, ...transformOptions } = this.options;

    const assets = compilation.getAssets().filter(
      (asset) =>
        // Filter out already minimized
        !asset.info.minimized &&
        // Filter out by file type
        CSS_REGEX.test(asset.name),
    );

    await Promise.all(
      assets.map(async (asset) => {
        const { source, map } = asset.source.sourceAndMap();
        const sourceAsString = source.toString();
        const code = typeof source === 'string' ? Buffer.from(source) : source;

        const result = this.transform({
          filename: asset.name,
          code,
          minify: true,
          sourceMap: sourcemap,
          ...transformOptions,
        });
        const codeString = result.code.toString();

        compilation.updateAsset(
          asset.name,
          // @ts-ignore
          sourcemap
            ? new SourceMapSource(
                codeString,
                asset.name,
                JSON.parse(result.map!.toString()),
                sourceAsString,
                map as any,
                true,
              )
            : new RawSource(codeString),
          {
            ...asset.info,
            minimized: true,
          },
        );
      }),
    );
  }
}

export default LightningCSSMinifyPlugin;
