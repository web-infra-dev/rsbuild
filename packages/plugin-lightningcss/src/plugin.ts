import path from 'node:path';
import type {
  ChainIdentifier,
  EnvironmentContext,
  RsbuildPlugin,
  RspackChain,
} from '@rsbuild/core';
import { reduceConfigs } from '@rsbuild/core';
import browserslist from '@rsbuild/shared/browserslist';
import * as lightningcss from 'lightningcss';
import type { Targets } from 'lightningcss';
import type {
  LightningCSSLoaderOptions,
  LightningCSSTransformOptions,
  Lightningcss,
  PluginLightningcssOptions,
} from './types';

export const PLUGIN_LIGHTNINGCSS_NAME = 'rsbuild:lightningcss';

const getLightningCSSTargets = async ({
  options,
  environment,
}: {
  options: PluginLightningcssOptions;
  environment: EnvironmentContext;
}) => {
  const browserslistUserConfig = environment.browserslist;
  const implementation =
    (options.implementation as Lightningcss) ?? lightningcss;

  const targets = implementation.browserslistToTargets(
    browserslist(browserslistUserConfig),
  );

  return targets;
};

const applyLightningCSSLoader = ({
  chain,
  CHAIN_ID,
  targets,
  options = {},
}: {
  chain: RspackChain;
  CHAIN_ID: ChainIdentifier;
  targets: Targets;
  options: PluginLightningcssOptions | undefined;
}) => {
  if (!chain.module.rules.has(CHAIN_ID.RULE.CSS)) {
    return;
  }

  const defaultOptions = {
    targets,
  } satisfies LightningCSSLoaderOptions;

  const { implementation } = options;

  if (options.transform === true) {
    options.transform = {};
  }

  const mergedOptions = reduceConfigs<LightningCSSTransformOptions>({
    initial: defaultOptions,
    config: {
      ...(implementation ? { implementation } : {}),
      ...options.transform,
    },
  });

  const rule = chain.module.rule(CHAIN_ID.RULE.CSS);

  rule
    .use(CHAIN_ID.USE.LIGHTNINGCSS)
    .loader(path.resolve(__dirname, './loader.cjs'))
    .options(mergedOptions)
    .after(CHAIN_ID.USE.CSS);

  rule.uses.delete(CHAIN_ID.USE.POSTCSS);
};

const validateImplementation = (implementation: unknown) => {
  if (!implementation) {
    return;
  }

  const lightningcss = implementation as Lightningcss;

  if (typeof implementation !== 'object') {
    throw new TypeError(
      `[${PLUGIN_LIGHTNINGCSS_NAME}]: implementation must be an 'lightningcss' object. Received ${typeof lightningcss}`,
    );
  }

  if (typeof lightningcss.transform !== 'function') {
    throw new TypeError(
      `[${PLUGIN_LIGHTNINGCSS_NAME}]: implementation.transform must be an 'lightningcss' transform function. Received ${typeof lightningcss.transform}`,
    );
  }

  if (typeof lightningcss.browserslistToTargets !== 'function') {
    throw new TypeError(
      `[${PLUGIN_LIGHTNINGCSS_NAME}]: options.implementation.browserslistToTargets must be an 'lightningcss' browserslistToTargets function. Received ${typeof lightningcss.browserslistToTargets}`,
    );
  }
};

export const pluginLightningcss = (
  options: PluginLightningcssOptions = {},
): RsbuildPlugin => ({
  name: PLUGIN_LIGHTNINGCSS_NAME,

  setup(api) {
    let targets: Targets;

    if (options.implementation) {
      validateImplementation(options.implementation);
    }

    api.modifyBundlerChain({
      // ensure lightningcss-loader can be applied to sass/less/stylus rules
      order: 'pre',

      handler: async (chain, { CHAIN_ID, target, environment }) => {
        targets = await getLightningCSSTargets({
          options,
          environment,
        });

        if (target === 'web' && options?.transform !== false) {
          applyLightningCSSLoader({
            chain,
            CHAIN_ID,
            targets,
            options,
          });
        }
      },
    });
  },
});
