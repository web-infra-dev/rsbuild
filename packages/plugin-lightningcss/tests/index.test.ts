import { createRsbuild } from '@rsbuild/core';
import {
  type RspackConfig,
  type RuleSetRule,
  isPlainObject,
} from '@rsbuild/shared';
import { createStubRsbuild } from '@scripts/test-helper';
import { describe, expect, it } from 'vitest';
import { lightningcss, pluginLightningcss } from '../src';

const getCssRules = (rspackConfig: RspackConfig) => {
  const CSS_RULES = ['css', 'scss', 'sass', 'less', 'stylus'];

  return (
    (rspackConfig.module?.rules?.filter((item) => {
      const isRule =
        isPlainObject(item) &&
        CSS_RULES.some((txt) =>
          item.test?.toString().replace(/[\W]/g, '').includes(txt),
        );
      return isRule;
    }) as RuleSetRule[]) ?? []
  );
};

describe('plugins/lightningcss', () => {
  it('plugin-lightningcss should replace postcss-loader with lightningcss-loader with default options', async () => {
    const rsbuild = await createRsbuild({
      rsbuildConfig: {
        plugins: [pluginLightningcss()],
      },
    });

    const bundlerConfigs = await rsbuild.initConfigs();
    const cssRules = getCssRules(bundlerConfigs[0]);
    expect(cssRules).toMatchSnapshot();
    expect(cssRules).not.contain('postcss-loader');
  });

  it('plugin-lightningcss should set lightningCssMinifyPlugin with default options', async () => {
    process.env.NODE_ENV = 'production';
    const rsbuild = await createStubRsbuild({
      plugins: [pluginLightningcss()],
    });
    const config = await rsbuild.unwrapConfig();
    expect(config.optimization).toMatchSnapshot();

    process.env.NODE_ENV = 'test';
  });

  it('plugin-lightningcss should not set lightningCssMinifyPlugin with minify: false', async () => {
    process.env.NODE_ENV = 'production';
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {
        output: {
          minify: false,
        },
      },
      plugins: [
        pluginLightningcss({
          minify: {
            errorRecovery: true,
            exclude: lightningcss.Features.Colors,
          },
        }),
      ],
    });

    const config = await rsbuild.unwrapConfig();
    expect(config.optimization).toMatchSnapshot();

    process.env.NODE_ENV = 'test';
  });

  it('plugin-lightningcss should not set lightningCssMinifyPlugin with minify.css: false', async () => {
    process.env.NODE_ENV = 'production';
    const rsbuild = await createStubRsbuild({
      rsbuildConfig: {
        output: {
          minify: {
            css: false,
          },
        },
      },
      plugins: [
        pluginLightningcss({
          minify: {
            errorRecovery: true,
            exclude: lightningcss.Features.Colors,
          },
        }),
      ],
    });

    const config = await rsbuild.unwrapConfig();
    expect(config.optimization).toMatchSnapshot();

    process.env.NODE_ENV = 'test';
  });

  it('plugin-lightningcss should be configurable by users', async () => {
    process.env.NODE_ENV = 'production';

    const rsbuild = await createRsbuild({
      rsbuildConfig: {
        plugins: [
          pluginLightningcss({
            implementation: {
              transform: lightningcss.transform,
              browserslistToTargets: lightningcss.browserslistToTargets,
            },
            transform: {
              errorRecovery: true,
              cssModules: {
                dashedIdents: true,
                pattern: '[hash]-[local]',
              },
              visitor: {
                Length(len) {
                  return {
                    unit: 'rem',
                    value: len.value,
                  };
                },
              },
            },
            minify: {
              errorRecovery: true,
              exclude: lightningcss.Features.Colors,
              cssModules: {
                dashedIdents: true,
                pattern: '[hash]-[local]',
              },
            },
          }),
        ],
      },
    });

    const bundlerConfigs = await rsbuild.initConfigs();
    const cssRules = getCssRules(bundlerConfigs[0]);
    expect(cssRules).toMatchSnapshot();
    expect(cssRules).not.contain('postcss-loader');

    expect(bundlerConfigs[0].optimization?.minimizer).toMatchSnapshot();

    process.env.NODE_ENV = 'test';
  });

  it('plugin-lightningcss should be configurable by users with true options', async () => {
    process.env.NODE_ENV = 'production';
    const rsbuild = await createRsbuild({
      rsbuildConfig: {
        plugins: [
          pluginLightningcss({
            transform: true,
            minify: true,
          }),
        ],
      },
    });

    const bundlerConfigs = await rsbuild.initConfigs();
    const cssRules = getCssRules(bundlerConfigs[0]);
    expect(cssRules).toMatchSnapshot();
    expect(cssRules).not.contain('postcss-loader');

    expect(bundlerConfigs[0].optimization?.minimizer).toMatchSnapshot();

    process.env.NODE_ENV = 'test';
  });

  it('plugin-lightningcss should be cancelable by users with false options', async () => {
    process.env.NODE_ENV = 'production';
    const rsbuild = await createStubRsbuild({
      plugins: [
        pluginLightningcss({
          transform: false,
          minify: false,
        }),
      ],
    });

    const bundlerConfigs = await rsbuild.initConfigs();
    expect(bundlerConfigs[0]).toMatchSnapshot();

    process.env.NODE_ENV = 'test';
  });
});
