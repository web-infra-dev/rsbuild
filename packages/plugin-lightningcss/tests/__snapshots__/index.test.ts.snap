// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugins/lightningcss > plugin-lightningcss should be cancelable by users with false options 1`] = `{}`;

exports[`plugins/lightningcss > plugin-lightningcss should be configurable by users 1`] = `
[
  {
    "resolve": {
      "preferRelative": true,
    },
    "sideEffects": true,
    "test": /\\\\\\.css\\$/,
    "use": [
      {
        "loader": "<ROOT>/node_modules/<PNPM_INNER>/@rspack/core/dist/builtin-plugin/css-extract/loader.js",
      },
      {
        "loader": "<ROOT>/packages/core/compiled/css-loader",
        "options": {
          "importLoaders": 1,
          "modules": {
            "auto": true,
            "exportGlobals": false,
            "exportLocalsConvention": "camelCase",
            "localIdentName": "[local]-[hash:base64:6]",
            "namedExport": false,
          },
          "sourceMap": false,
        },
      },
      {
        "loader": "<ROOT>/packages/plugin-lightningcss/src/loader.cjs",
        "options": {
          "cssModules": {
            "dashedIdents": true,
            "pattern": "[hash]-[local]",
          },
          "errorRecovery": true,
          "implementation": {
            "browserslistToTargets": [Function],
            "transform": [Function],
          },
          "targets": {
            "chrome": 5701632,
            "edge": 5767168,
            "firefox": 5111808,
            "safari": 917504,
          },
          "visitor": {
            "Length": [Function],
          },
        },
      },
    ],
  },
]
`;

exports[`plugins/lightningcss > plugin-lightningcss should be configurable by users 2`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_options": {
      "compress": "{"passes":1,"pure_funcs":[],"drop_console":false}",
      "exclude": undefined,
      "extractComments": {
        "condition": "@preserve|@lic|@cc_on|^/**!",
      },
      "format": "{"comments":false,"asciiOnly":true}",
      "include": undefined,
      "mangle": "{"keep_classnames":false,"keep_fnames":false}",
      "module": undefined,
      "test": undefined,
    },
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  LightningCSSMinifyPlugin {
    "options": {
      "cssModules": {
        "dashedIdents": true,
        "pattern": "[hash]-[local]",
      },
      "errorRecovery": true,
      "exclude": 64512,
      "implementation": {
        "browserslistToTargets": [Function],
        "transform": [Function],
      },
      "targets": {
        "chrome": 5701632,
        "edge": 5767168,
        "firefox": 5111808,
        "safari": 917504,
      },
    },
    "transform": [Function],
  },
]
`;

exports[`plugins/lightningcss > plugin-lightningcss should be configurable by users with true options 1`] = `
[
  {
    "resolve": {
      "preferRelative": true,
    },
    "sideEffects": true,
    "test": /\\\\\\.css\\$/,
    "use": [
      {
        "loader": "<ROOT>/node_modules/<PNPM_INNER>/@rspack/core/dist/builtin-plugin/css-extract/loader.js",
      },
      {
        "loader": "<ROOT>/packages/core/compiled/css-loader",
        "options": {
          "importLoaders": 1,
          "modules": {
            "auto": true,
            "exportGlobals": false,
            "exportLocalsConvention": "camelCase",
            "localIdentName": "[local]-[hash:base64:6]",
            "namedExport": false,
          },
          "sourceMap": false,
        },
      },
      {
        "loader": "<ROOT>/packages/plugin-lightningcss/src/loader.cjs",
        "options": {
          "targets": {
            "chrome": 5701632,
            "edge": 5767168,
            "firefox": 5111808,
            "safari": 917504,
          },
        },
      },
    ],
  },
]
`;

exports[`plugins/lightningcss > plugin-lightningcss should be configurable by users with true options 2`] = `
[
  SwcJsMinimizerRspackPlugin {
    "_options": {
      "compress": "{"passes":1,"pure_funcs":[],"drop_console":false}",
      "exclude": undefined,
      "extractComments": {
        "condition": "@preserve|@lic|@cc_on|^/**!",
      },
      "format": "{"comments":false,"asciiOnly":true}",
      "include": undefined,
      "mangle": "{"keep_classnames":false,"keep_fnames":false}",
      "module": undefined,
      "test": undefined,
    },
    "affectedHooks": "compilation",
    "name": "SwcJsMinimizerRspackPlugin",
  },
  LightningCSSMinifyPlugin {
    "options": {
      "targets": {
        "chrome": 5701632,
        "edge": 5767168,
        "firefox": 5111808,
        "safari": 917504,
      },
    },
    "transform": [Function],
  },
]
`;

exports[`plugins/lightningcss > plugin-lightningcss should not set lightningCssMinifyPlugin with minify.css: false 1`] = `undefined`;

exports[`plugins/lightningcss > plugin-lightningcss should not set lightningCssMinifyPlugin with minify: false 1`] = `undefined`;

exports[`plugins/lightningcss > plugin-lightningcss should replace postcss-loader with lightningcss-loader with default options 1`] = `
[
  {
    "resolve": {
      "preferRelative": true,
    },
    "sideEffects": true,
    "test": /\\\\\\.css\\$/,
    "use": [
      {
        "loader": "<ROOT>/node_modules/<PNPM_INNER>/@rspack/core/dist/builtin-plugin/css-extract/loader.js",
      },
      {
        "loader": "<ROOT>/packages/core/compiled/css-loader",
        "options": {
          "importLoaders": 1,
          "modules": {
            "auto": true,
            "exportGlobals": false,
            "exportLocalsConvention": "camelCase",
            "localIdentName": "[path][name]__[local]-[hash:base64:6]",
            "namedExport": false,
          },
          "sourceMap": false,
        },
      },
      {
        "loader": "<ROOT>/packages/plugin-lightningcss/src/loader.cjs",
        "options": {
          "targets": {
            "chrome": 5701632,
            "edge": 5767168,
            "firefox": 5111808,
            "safari": 917504,
          },
        },
      },
    ],
  },
]
`;

exports[`plugins/lightningcss > plugin-lightningcss should set lightningCssMinifyPlugin with default options 1`] = `
{
  "minimizer": [
    LightningCSSMinifyPlugin {
      "options": {
        "targets": {
          "chrome": 5701632,
          "edge": 5767168,
          "firefox": 5111808,
          "safari": 917504,
        },
      },
      "transform": [Function],
    },
  ],
}
`;
